server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -351,6 +354,26 @@ public TimeValue timeout() {
         }, listener::onFailure);
     }
 
+    public boolean hasOldVersionSnapshots(String repositoryName, RepositoryData repositoryData, @Nullable SnapshotId excluded) {
+        final Collection<SnapshotId> snapshotIds = repositoryData.getSnapshotIds();
+        final boolean hasOldFormatSnapshots;
+        if (snapshotIds.isEmpty()) {
+            hasOldFormatSnapshots = false;
+        } else {
+            if (repositoryData.shardGenerations().totalShards() > 0) {
+                hasOldFormatSnapshots = false;
+            } else {
+                final Repository repository = repositoriesService.repository(repositoryName);
+                hasOldFormatSnapshots = snapshotIds.stream().map(repository::getSnapshotInfo).anyMatch(
```

Comment:
what if we can't load one of these snapshot info? Should we be lenient in that case? Will this o.w. break snapshotting in a way that wasn't broken before?

Reply:
This will brake snapshotting and deleting snapshots if a `snap-uuid` file doesn't exist.
I think it's much better to fail this case than to quietly ignore it. 
We simply should definitely not be making any updates to a corrupted repo ever IMO. This will not break restoring snapshots from the repo and users should simply move to a new repo when running into this kind of corruption. We discussed this before and I think the conclusion was that if a `snap-` file is missing ... what guarantees do we have some `__` data file isn't missing as well and quietly causing any new snapshot that uses it to be corrupted as well.
We also already eliminated a number of other spots around repo corruption like this where we were lenient before but aren't any longer.

Reply:
I'm mainly concerned about the situation in Cloud where repos that got previously broken due to eventual consistencies are not taking any new backups anymore, and require an operator to come in and fix the situation. I have a strong preference here to add special handling for the case where the meta file is not found and be lenient in that case (and assume it's an old style repo).

Reply:
> I'm mainly concerned about the situation in Cloud where repos that got previously broken due to eventual consistencies are not taking any new backups anymore, and require an operator to come in and fix the situation.

I would argue that this is what we want. Assuming you have a snapshot in the `index-N` for which the `snap-` blob is missing at the repo root, then what guarantees do you have that the data blobs belonging to that snapshot and are referenced in the shard level `index-N` do actually exist and you're not just creating and endless stream of corrupted snapshots quietly? IMO this situation should be resolved in Cloud operationally (and if it turns out to be common, automated).

Reply:
> Assuming you have a snapshot in the index-N for which the snap- blob is missing at the repo root, then what guarantees do you have that the data blobs belonging to that snapshot and are referenced in the shard level index-N do actually exist and you're not just creating and endless stream of corrupted snapshots quietly?

I don't see how these two are related. Having a missing top-level `snap-` blob does not entail that you have missing segment blobs. I therefore don't follow the argument for endless corrupted snapshots. In any case, I don't think it's the responsibility of the snapshot creation / snapshot deletion API to verify unrelated snapshots, which is what this API now suddenly does with the changes here as it happens to need a full top-level scan of the repo. Let's leave this kind of verification to a separate API that checks integrity of the repo.

Reply:
> I don't see how these two are related. Having a missing top-level snap- blob does not entail that you have missing segment blobs. I therefore don't follow the argument for endless corrupted snapshots.

Two points here:
1. The argument isn't specific to S3 here. The absence of a file that is supposed to be there but is not means that the repository is corrupted and other files could be missing for whatever reason (say a broken NFS setup).
2. Concretely to the point, If a `snap-` blob is missing due to eventual consistency issues the sequence of events would be:
  * `snap-` and all the other meta and data blobs related to a snapshot were deleted and a new index-N was written
  * Next operation failed to pick up `index-N` due to inconsistent listing and instead picked up `index-(N-1)` that still included the snapshot and then writes out `index-N` again still including the snapshot.
  * If the two operations were so close to each other, maybe we also picked up an older shard level `index-N` blob that references data blobs that are actually gone already? I don't think this can be ruled out at all and is somewhat likely even.

>  I don't think it's the responsibility of the snapshot creation / snapshot deletion API to verify unrelated snapshots, which is what this API now suddenly does with the changes here as it happens to need a full top-level scan of the repo. Let's leave this kind of verification to a separate API that checks integrity of the repo.

Agreed, I don't think it's the responsibility of snapshot create and delete to do any verification outright. But I though consensus was that if either of these operations detected the repository to be in a corrupted state, it wouldn't continue to make updates to the repo. If I remember correctly, the reasoning was precisely that if e.g. an NFS issue makes some files go missing and we continue to modify the repo ignoring that corruption, then we take away any chance of the operator to fix the storage layer instead of losing those snapshots for good?

Reply:
@ywelsch as discussed on another channel, I adjusted things to now ignored a missing `snap-${uuid}.dat` blob and assume old generation in that case.
I also added a test for this scenario in 7d5d2565b0e (I know it's a little hacky, but I still liked it better than starting to manually mess with files in the repos used by the REST/BwC tests).